consumer/consumererror/networkerrors.go

```diff
@@ -0,0 +1,89 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
+
+import (
+	"fmt"
+	"net/http"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer/consumererror/internal/statusconversion"
+)
+
+// NetworkError contains either an HTTP or gRPC status code
+// resulting from a failed synchronous network request.
+// It allows retrieving a type of error code, converting
+// between the status codes supported by each transport if necessary.
+//
+// It should be created with NewHTTPStatus or NewGRPCStatus.
+type NetworkError struct {
+	error
+	httpStatus *int
+	grpcStatus *status.Status
+}
+
+var _ error = NetworkError{}
+
+// Error returns a string specifying the transport and corresponding status code.
+func (se NetworkError) Error() string {
+	if se.httpStatus != nil {
+		return fmt.Sprintf("HTTP Status (%d): %s", *se.httpStatus, se.error.Error())
+	} else if se.grpcStatus != nil {
+		return fmt.Sprintf("gRPC Status (%s): %s", se.grpcStatus.Code().String(), se.error.Error())
+	}
+
+	return "Network error (no error code set): " + se.error.Error()
```

Comment:
The only place this would be unset if someone tried to create it outside the New* funcs. If we want to unexport it we need to make a `consumererror.AsNetworkError` function or similar to pull out a `NetworkError` interface that we control.

I wanted to allow users to just use the stdlib `errors.As` function to make these errors feel more idiomatic and to avoid the extra implementation overhead, but if we're concerned about users circumventing the New* functions we can hide the error type.